src/stats/cdf.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2016 Johannes Köster.
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Support for discrete probability distributions in terms of cumulative distribution
+//! functions (CDF).
+
+use std::f64;
+use std::iter;
+use std::slice;
+
+use num::traits::{cast, NumCast};
+use itertools::Itertools;
+
+use stats::logprobs::{self, LogProb};
+
+
+/// Implementation of a cumulative distribution function.
+#[derive(Debug, Clone)]
+pub struct CDF<T: PartialOrd> {
+    inner: Vec<(T, LogProb)>
+}
+
+
+impl<T: PartialOrd> CDF<T> {
+    /// Create CDF from given probability mass function (PMF). The PMF may contain duplicate values
+    /// the probabilities of which are summed during generation of the CDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `pmf` - the PMF as a vector of value/probability pairs
+    pub fn from_pmf(mut entries: Vec<(T, LogProb)>) -> Self {
+        entries.sort_by(|&(ref a, _), &(ref b, _)| a.partial_cmp(b).unwrap());
+        let mut inner: Vec<(T, LogProb)> = Vec::new();
+        for mut e in entries.into_iter() {
+            let p = logprobs::add(inner.last().map_or(f64::NEG_INFINITY, |e| e.1), e.1);
+            if !inner.is_empty() && inner.last().unwrap().0 == e.0 {
+                inner.last_mut().unwrap().1 = p;
+            }
+            else {
+                e.1 = p;
+                inner.push(e);
+            }
+        }
+        let mut cdf = CDF {
+            inner: inner
+        };
+
+        if relative_eq!(cdf.total_prob(), 0.0) && cdf.total_prob() > 0.0 {
```

Comment:
I think I understand what you're doing here, but it reads rather oddly: "If this value is pretty much zero and it's greater than zero..." I think this would make more sense as an expression checking if `total_prob` is greater than zero and less than some lambda.